From bc952df11e4f9ebcb511fbbcf88d80cbfcb799f7 Mon Sep 17 00:00:00 2001
From: Subrata Chatterjee <schatterjee@habana.ai>
Date: Wed, 14 Oct 2020 19:48:04 +0300
Subject: [PATCH] [SW-21652] goya2: ppboot: halt and reset arcs

In legacy bootflow ARM boots first and then it boots
ARCs. ARCs need to be explicitly restarted for cases where
those are not done by hardware. For example when LKD module(on host)
is removed (rmmod), LKD resets ARM but ARCs continue to run.
So ARCs need to be restarted by ARM in the boot flow.
To cater this flow ARCs are first halted and then triggered for reset.

Change-Id: I071a31cae1b4476176225b23fe278d885f592c01
Signed-off-by: Subrata Chatterjee <schatterjee@habana.ai>
---

diff --git a/goya2/common.c b/goya2/common.c
index 72d7efc..3495869 100644
--- a/goya2/common.c
+++ b/goya2/common.c
@@ -217,6 +217,9 @@
 
 static void init_arc(void)
 {
+	HALT_N_RESET_ARC(0);
+	HALT_N_RESET_ARC(1);
+
 	/*
 	 * Program Reset vector address for ARC0 and ARC1 to ICCM 0x0
 	 * location where vector table first entry will be available and
diff --git a/goya2/common.h b/goya2/common.h
index 0400687..6458a71 100644
--- a/goya2/common.h
+++ b/goya2/common.h
@@ -20,3 +20,16 @@
 
 #define ARC_RST_VALUE			0x110
 #define ARC_RST_VEC_ADDR		0x0
+
+/*
+ * Set ARCs in halt state before reset
+ * 1. Set halt bit in RUN_HALT_REQ register
+ * 2. Clear halt bit in RUN_HALT_REQ register
+ * 3. Set Reset bit
+ */
+#define HALT_N_RESET_ARC(id)						 \
+	do {								 \
+		regmap_writel(BIT(1), mmPSOC_ARC##id##_CFG_RUN_HALT_REQ);\
+		regmap_writel(0x0, mmPSOC_ARC##id##_CFG_RUN_HALT_REQ);	 \
+		regmap_writel(0x1, mmPSOC_ARC##id##_CFG_ARC_RST);	 \
+	} while (0)
